### PR TITLE
STITCH-872 Create Client API v2.0 for Android SDK

### PR DESCRIPTION
**Pull request by jsflax**

### CODE REVIEW

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientServiceTest.kt [hunk 1]
```diff
@@ -24,12 +27,13 @@ class StitchClientServiceTest {
     private val stitchClient: StitchClient by lazy {
         StitchClient(
                 instrumentationCtx,
-                "test-uybga"
+                "android-zgeec",
+                "https://stitch-qa.mongodb.com"
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientServiceTest.kt]: remove this

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientServiceTest.kt [hunk 2]
```diff
@@ -50,6 +54,12 @@ class StitchClientServiceTest {
         }
     }
 
+    @Test
+    fun testGetAuthProviders() {
+        val providers = await(this.stitchClient.authProviders)
+        print(providers)
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientServiceTest.kt]: needed

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientServiceTest.kt [hunk 3]
```diff
@@ -50,6 +54,12 @@ class StitchClientServiceTest {
         }
     }
 
+    @Test
+    fun testGetAuthProviders() {
+        val providers = await(this.stitchClient.authProviders)
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientServiceTest.kt]: need the val?

File: sdk/src/main/java/com/mongodb/stitch/android/CustomBsonConverter.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import org.bson.Document;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Created by jasonflax on 11/21/17.
```
Comment (edaniels) [inline sdk/src/main/java/com/mongodb/stitch/android/CustomBsonConverter.java]: remove

File: sdk/src/main/java/com/mongodb/stitch/android/CustomBsonConverter.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+ */
+
+public class CustomBsonConverter {
```
Comment (edaniels) [inline sdk/src/main/java/com/mongodb/stitch/android/CustomBsonConverter.java]: add depth

File: sdk/src/main/java/com/mongodb/stitch/android/CustomBsonConverter.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                throw new BsonInvalidOperationException(
+                        entry + " was not formatted properly");
+            }
+        }
+
+        return new BsonDocument(bsonElements);
+    }
+
+    public static BsonValue fromObject(Object arg) {
+        if (arg == null) {
```
Comment (edaniels) [inline sdk/src/main/java/com/mongodb/stitch/android/CustomBsonConverter.java]: groovy

File: sdk/src/main/java/com/mongodb/stitch/android/CustomObjectMapper.java [hunk 7]
```diff
@@ -17,11 +17,11 @@
  * CustomObjectMapper is responsible for handling the serialization and deserialization of JSON
  * objects with special serialization support for {@link Document}s and {@link ObjectId}s
  */
-class CustomObjectMapper {
+public class CustomObjectMapper {
```
Comment (edaniels) [inline sdk/src/main/java/com/mongodb/stitch/android/CustomObjectMapper.java]: Why did this become public?

File: sdk/src/main/java/com/mongodb/stitch/android/StitchClient.java [hunk 8]
```diff
@@ -604,73 +606,65 @@ public void onErrorResponse(final VolleyError error) {
         return future.getTask();
     }
 
-    // Pipelines
+    public Task<Object> executeFunction(String name, BsonValue... args) {
```
Comment (edaniels) [inline sdk/src/main/java/com/mongodb/stitch/android/StitchClient.java]: This takes BsonValues but the other takes objects

File: sdk/src/main/java/com/mongodb/stitch/android/StitchClient.java [hunk 9]
```diff
@@ -604,73 +606,65 @@ public void onErrorResponse(final VolleyError error) {
...
-    @SuppressWarnings("unchecked")
-    public Task<List<Object>> executePipeline(final List<PipelineStage> pipeline) {
+    public Task<Object> executeServiceFunction(String name, String serviceName, Object... args) {
         ensureAuthenticated();
-        final String pipeStr;
-        try {
-            pipeStr = _objMapper.writeValueAsString(pipeline);
-        } catch (final IOException e) {
-            return Tasks.forException(e);
+        Document doc = new Document("name", name);
```
Comment (edaniels) [inline sdk/src/main/java/com/mongodb/stitch/android/StitchClient.java]: final

File: sdk/src/main/java/com/mongodb/stitch/android/StitchClient.java [hunk 10]
```diff
@@ -604,73 +606,65 @@ public void onErrorResponse(final VolleyError error) {
...
+        doc.put("arguments", CustomBsonConverter.fromArray(args));
+        if (serviceName != null) {
+            doc.put("service", serviceName);
         }
 
-        return executeRequest(Request.Method.POST, Paths.PIPELINE, pipeStr).continueWith(new Continuation<String, List<Object>>() {
+        String json = doc.toJson();
+        System.out.println(json);
```
Comment (edaniels) [inline sdk/src/main/java/com/mongodb/stitch/android/StitchClient.java]: remove

File: sdk/src/main/java/com/mongodb/stitch/android/StitchClient.java [hunk 11]
```diff
@@ -539,24 +545,20 @@ public synchronized void removeAuthListener(final AuthListener authListener) {
...
+        final JsonArrayRequest request = new JsonArrayRequest(
                 Request.Method.GET,
                 url,
-                new Response.Listener<JSONObject>() {
+                new Response.Listener<JSONArray>() {
                     @Override
-                    public void onResponse(final JSONObject response) {
-
+                    public void onResponse(final JSONArray response) {
+                        System.out.println(response);
```
Comment (edaniels) [inline sdk/src/main/java/com/mongodb/stitch/android/StitchClient.java]: remove any printlns

File: sdk/src/main/java/com/mongodb/stitch/android/auth/AuthProviderInfo.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * Created by jasonflax on 11/20/17.
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+public abstract class AuthProviderInfo {
+    public static class Fields {
+        public static final String ID = "_id";
```
Comment (edaniels) [inline sdk/src/main/java/com/mongodb/stitch/android/auth/AuthProviderInfo.java]: Same removals here as ios

File: sdk/src/main/java/com/mongodb/stitch/android/auth/emailpass/EmailPasswordAuthProvider.java [hunk 13]
```diff
@@ -11,8 +11,8 @@
  * EmailPasswordAuthProvider provides a way to authenticate using an email and password.
  */
 public class EmailPasswordAuthProvider implements AuthProvider {
-    private static final String AUTH_TYPE = "local";
-    private static final String AUTH_NAME = "userpass";
+    private static final String AUTH_TYPE = "local-userpass";
+    private static final String AUTH_NAME = "local-userpass";
```
Comment (edaniels) [inline sdk/src/main/java/com/mongodb/stitch/android/auth/emailpass/EmailPasswordAuthProvider.java]: No more auth_names

File: sdk/src/main/java/com/mongodb/stitch/android/auth/oauth2/facebook/FacebookAuthProviderInfo.java [hunk 14]
```diff
@@ -1,42 +1,96 @@
...
+        }
     }
 
-    public String getApplicationId() {
-        return _appId;
+    private List<MetadataField> _metadataFields;
+    private Config _config;
+
+    @JsonCreator
+    public FacebookAuthProviderInfo(@JsonProperty(AuthProviderInfo.Fields.ID) @NonNull final String id,
```
Comment (edaniels) [inline sdk/src/main/java/com/mongodb/stitch/android/auth/oauth2/facebook/FacebookAuthProviderInfo.java]: Will have to make same removals here for fields not there

File: sdk/src/main/java/com/mongodb/stitch/android/CustomBsonConverter.java [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Created by jasonflax on 11/21/17.
+ */
+
+public class CustomBsonConverter {
+    public static BsonArray fromArray(Object[] array) {
+        BsonArray bsonArray = new BsonArray();
```
Comment (haleyowen) [inline sdk/src/main/java/com/mongodb/stitch/android/CustomBsonConverter.java]: Idk how big we are on finals, but it seems like everywhere else we do it, so maybe add them? I'm only commenting on it here but could be applied to whole PR. :)

File: sdk/src/main/java/com/mongodb/stitch/android/auth/AuthProviderInfo.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * Created by jasonflax on 11/20/17.
+ */
```
Comment (haleyowen) [inline sdk/src/main/java/com/mongodb/stitch/android/auth/AuthProviderInfo.java]: remove this

File: sdk/src/main/java/com/mongodb/stitch/android/auth/emailpass/EmailPasswordAuthProviderInfo.java [hunk 17]
```diff
@@ -1,12 +1,66 @@
...
+        static class Fields {
+            static final String EMAIL_CONFIRMATION_URL = "emailConfirmationUrl";
+            static final String RESET_PASSWORD_URL = "resetPasswordUrl";
+        }
+
+        private String _emailConfirmationUrl;
+        private String _resetPasswordUrl;
+
+        public Config(@JsonProperty(Fields.EMAIL_CONFIRMATION_URL)
+                      @NonNull String emailConfirmationUrl,
```
Comment (haleyowen) [inline sdk/src/main/java/com/mongodb/stitch/android/auth/emailpass/EmailPasswordAuthProviderInfo.java]: Put these on the same line

File: sdk/src/main/java/com/mongodb/stitch/android/auth/oauth2/facebook/FacebookAuthProviderInfo.java [hunk 18]
```diff
@@ -1,42 +1,96 @@
...
+            static final String REQUIRED = "required";
+        }
+
+        private String _name;
+        private boolean _required;
+
+        private MetadataField(@JsonProperty(Fields.NAME)
+                              @NonNull String name,
+                              @JsonProperty(Fields.REQUIRED)
+                                      boolean required) {
```
Comment (haleyowen) [inline sdk/src/main/java/com/mongodb/stitch/android/auth/oauth2/facebook/FacebookAuthProviderInfo.java]: same here

File: sdk/src/main/java/com/mongodb/stitch/android/auth/oauth2/facebook/FacebookAuthProviderInfo.java [hunk 19]
```diff
@@ -1,42 +1,96 @@
...
+        @NonNull
+        public boolean getRequired() {
+            return this._required;
+        }
+    }
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    public static class Config {
+        static class Fields {
+            static final String CLIENT_ID = "clientId";
```
Comment (haleyowen) [inline sdk/src/main/java/com/mongodb/stitch/android/auth/oauth2/facebook/FacebookAuthProviderInfo.java]: Should this be "client_id"?

File: sdk/src/main/java/com/mongodb/stitch/android/auth/oauth2/google/GoogleAuthProviderInfo.java [hunk 20]
```diff
@@ -1,54 +1,101 @@
...
         }
+    }
 
-        final List<Scope> gScopes = new ArrayList<>(scopes.size());
-        for (final String scope : scopes) {
-            gScopes.add(new Scope(scope));
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    public static class Config {
+        static class Fields {
+            static final String CLIENT_ID = "clientId";
```
Comment (haleyowen) [inline sdk/src/main/java/com/mongodb/stitch/android/auth/oauth2/google/GoogleAuthProviderInfo.java]: here too?

File: sdk/src/main/java/com/mongodb/stitch/android/push/PushClient.java [hunk 21]
```diff
@@ -13,13 +15,24 @@
  * sent from Stitch or directly from the provider.
  */
 public abstract class PushClient {
+    protected final class Routes {
+        public final String PUSH_PROVIDERS_ROUTE =
+                String.format("app/%s/push/providers", _stitchClient.getAppId());
+
+        public String getPushProvidersRegistartionRoute(String provider) {
```
Comment (haleyowen) [inline sdk/src/main/java/com/mongodb/stitch/android/push/PushClient.java]: Registration is spelt wrong

File: sdk/src/main/java/com/mongodb/stitch/android/services/mongodb/MongoClient.java [hunk 22]
```diff
@@ -247,54 +110,78 @@ public PipelineStage makeDeleteStage(
...
+            return _database._client._stitchClient.executeServiceFunction(
+                "find", _database._client._service, doc
+            ).continueWith(new Continuation<Object, List<Document>>() {
+                @Override
+                public List<Document> then(@NonNull Task<Object> task) throws Exception {
+                    if (task.isSuccessful()) {
+                        Object result = task.getResult();
+                        if (result instanceof JSONArray) {
+                            List<Document> docs = new ArrayList<>();
+                            for (int i = 0; i < ((JSONArray) result).length(); i++) {
```
Comment (haleyowen) [inline sdk/src/main/java/com/mongodb/stitch/android/services/mongodb/MongoClient.java]: Could probably use fancy for loop here

File: sdk/src/main/java/com/mongodb/stitch/android/services/mongodb/MongoClient.java [hunk 23]
```diff
@@ -356,19 +251,28 @@ public Void then(@NonNull final Task<List<Object>> task) throws Exception {
...
+            ).continueWith(new Continuation<Object, Document>() {
                 @Override
-                public Void then(@NonNull final Task<List<Object>> task) throws Exception {
+                public Document then(@NonNull Task<Object> task) throws Exception {
                     if (task.isSuccessful()) {
-                        return null;
+                        Object result = task.getResult();
+                        return Document.parse(result.toString());
+                    } else {
+                        Log.e(TAG, "Error while executing function", task.getException());
```
Comment (haleyowen) [inline sdk/src/main/java/com/mongodb/stitch/android/services/mongodb/MongoClient.java]: Did we not want to leave these error messages to be more specific?

Comment (jsflax) [reply sdk/src/main/java/com/mongodb/stitch/android/services/mongodb/MongoClient.java]: The exception should contain the more specific information.

File: sdk/src/main/java/com/mongodb/stitch/android/services/mongodb/MongoClient.java (hunk 22 above)
Comment (jsflax) [reply sdk/src/main/java/com/mongodb/stitch/android/services/mongodb/MongoClient.java]: foreach is not applicable with `JSONArray`

File: sdk/src/main/java/com/mongodb/stitch/android/push/PushClient.java (hunk 21 above)
Comment (jsflax) [reply sdk/src/main/java/com/mongodb/stitch/android/push/PushClient.java]: good catch

File: sdk/src/main/java/com/mongodb/stitch/android/auth/oauth2/google/GoogleAuthProviderInfo.java (hunk 20 above)
Comment (jsflax) [reply sdk/src/main/java/com/mongodb/stitch/android/auth/oauth2/google/GoogleAuthProviderInfo.java]: Unfortunately we didn't update these to snakecase on the serverside

File: sdk/src/main/java/com/mongodb/stitch/android/auth/oauth2/facebook/FacebookAuthProviderInfo.java (hunk 19 above)
Comment (jsflax) [reply sdk/src/main/java/com/mongodb/stitch/android/auth/oauth2/facebook/FacebookAuthProviderInfo.java]: Unfortunately we didn't update these to snakecase on the serverside

**Pull request merged**